### PR TITLE
periodically reload tokens read from TokenFile in kubeconfig

### PR DESCRIPTION
**Pull request by mikedanese**
Like we do with InClusterConfig.

/kind cleanup
/sig auth

```release-note
Go clients created from a kubeconfig that specifies a TokenFile now periodically reload the token from the specified file.
```

### CODE REVIEW

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go [hunk 1]
```diff
@@ -229,11 +229,12 @@ func (config *DirectClientConfig) getUserIdentificationPartialConfig(configAuthI
 	if len(configAuthInfo.Token) > 0 {
 		mergedConfig.BearerToken = configAuthInfo.Token
 	} else if len(configAuthInfo.TokenFile) > 0 {
-		tokenBytes, err := ioutil.ReadFile(configAuthInfo.TokenFile)
+		ts := restclient.NewCachedFileTokenSource(configAuthInfo.TokenFile)
+		_, err := ts.Token()
```
Comment (awly) [inline staging/src/k8s.io/client-go/tools/clientcmd/client_config.go]: merge this with error handling

**Review by awly — COMMENTED** [review]
LGTM, one nit

**Comment by mikedanese** [comment]
@awly Fixed, PTAL.

**Comment by mikedanese** [comment]
Kick the bot:
/approve

**Comment by roycaihw** [comment]
cc @roycaihw @mbohlool @yliaog

File: staging/src/k8s.io/client-go/tools/clientcmd/client_config.go [hunk 2]
```diff
@@ -229,11 +229,11 @@ func (config *DirectClientConfig) getUserIdentificationPartialConfig(configAuthI
 	if len(configAuthInfo.Token) > 0 {
 		mergedConfig.BearerToken = configAuthInfo.Token
 	} else if len(configAuthInfo.TokenFile) > 0 {
-		tokenBytes, err := ioutil.ReadFile(configAuthInfo.TokenFile)
-		if err != nil {
+		ts := restclient.NewCachedFileTokenSource(configAuthInfo.TokenFile)
```
Comment (smarterclayton) [inline staging/src/k8s.io/client-go/tools/clientcmd/client_config.go]: This broke clients that depend on these methods.  We shouldn't have broken this without providing an alternate path.

Comment (soltysh) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go]: I agree with Clayton here, `BearerToken` was never deprecated and we don't provide alternative, yet suddenly it's empty. We should deprecate this field and provide users with explanation what they should be using instead.

Comment (soltysh) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go]: > The transformation of TokenFile -> bearer token (and what InClusterConfig returns) is what is in question.

Yes, sorry for my short-circuited answer :)

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go]: I'm looking into plumbing both BearerToken and TokenFile out to the config and moving the refresh behavior into the round tripper constructed from the config

Comment (liggitt) [reply staging/src/k8s.io/client-go/tools/clientcmd/client_config.go]: opened https://github.com/kubernetes/kubernetes/pull/71713

**Pull request merged**